File: Python programs/linearsearchwiththreading.py
import threading
import time

stopthreads=0
print_lock = threading.Lock()
def linsearch1(list1,ele1):
    pos1 = -1
    n = len(list1)
    mid = int(((n+1)/2)-1)
    global stopthreads
    for i in range(0,mid+1):
        if(stopthreads==0):
           
            if list1[i]==ele1:
                pos1 = i
                stopthreads=1
                break
        else:
            break
    with print_lock:
        print("thread1 pos ",pos1)
 

def linsearch2(list2,ele2):
    pos2=-1
    global stopthreads
    n = len(list2)
    mid = int(((n+1)/2)-1)
    for i in range(mid,n):
        if(stopthreads==0):
            if list2[i]==ele2:
                pos2=i
                stopthreads=1
                break
        else:
            break
    with print_lock:
        print("thread2 pos ",pos2)

def linsearch3(list3,ele3):
    pos3=-1
    global stopthreads
    n = len(list3)
    mid = int(((n+1)/2)-1)
    for i in range(mid,-1,-1):
        if(stopthreads==0):
            if list3[i]==ele3:
                pos3=i
                stopthreads=1
                break
        else:
            break
    with print_lock:
        print("thread3 pos ",pos3)

def linsearch4(list4,ele4):
    pos4=-1
    global stopthreads
    n = len(list4)
    mid = int(((n+1)/2)-1)
    for i in range(n-1,mid-1,-1):
        if(stopthreads==0):
            if list4[i]==ele4:
                pos4=i
                stopthreads=1
                break
        else:
            break
    with print_lock:
        print("thread4 pos ",pos4)
    




n=int(input("how many elements?"))
list5 = []
for i in range(1,n+1):
    list5.append(i)
ele = int(input("enter the element to be searched"))


t1 = threading.Thread(target=linsearch1, args = (list5,ele))
t2 = threading.Thread(target=linsearch2, args = (list5,ele))
t3 = threading.Thread(target=linsearch3, args = (list5,ele))
t4 = threading.Thread(target=linsearch4, args = (list5,ele))

start = time.time()
    
t1.start()
t3.start()
t2.start()
t4.start()

t1.join()
t3.join()
t2.join()
t4.join()

print("Time taken ",time.time()-start)

File: C and C++ PROGRAMS/practical applications/linearsearchwiththreading.cpp
#include<thread>
#include<iostream>
#include<ctime>
#include<stdio.h>
#include<mutex>
#include<windows.h>
using namespace std;
long long n;
void linsearch1(int*,int);
void linsearch2(int*,int);
void linsearch3(int*,int);
void linsearch4(int*,int);
int stopthreads=0;
int main()
{
    int i,ele;
    double start;
    cout<<"Enter the number of elements"<<endl;
    cin>>n;
    int* a = new int[n];
    for(i=1;i<n+1;i++)
        a[i-1]=i;
    cout<<"Enter the element to be searched"<<endl;
    cin>>ele;
    start = time(0);
    thread t1(linsearch1,a,ele);
    thread t2(linsearch2,a,ele);
    thread t3(linsearch3,a,ele);
    thread t4(linsearch4,a,ele);
    t1.join();
    t2.join();
    t3.join();
    t4.join();
    cout<<endl<<time(0)-start;
    return 0;
}
void linsearch1(int* a1,int ele1)
{
    int pos1=-1,i,mid;
    mutex m;
    mid = int(((n+1)/2)-1);
    for(i=0;i<=mid;i++)
    {
        if(stopthreads==0)
        {

            if(a1[i]==ele1)
            {
                pos1=i;
                stopthreads=1;
                break;
            }
        }
        else
        {
         Sleep(100);
         break;
        }
    }
    m.lock();
    cout<<"thread1 pos1 "<<pos1<<endl;
    m.unlock();
}
void linsearch2(int* a2,int ele2)
{
    mutex m;
    int pos2=-1,i,mid;
    mid = int(((n+1)/2)-1);
    for(i=mid;i<n;i++)
    {
        if(stopthreads==0)
        {
            if(a2[i]==ele2)
            {
                pos2=i;
                stopthreads=1;
                break;
            }
        }
        else
        {
            Sleep(100);
            break;
        }
    }
    m.lock();
    cout<<"thread2 pos2 "<<pos2<<endl;
    m.unlock();
}
void linsearch3(int* a3,int ele3)
{
    mutex m;
    int pos3=-1,i,mid;
    mid = int(((n+1)/2)-1);
    for(i=mid;i>=0;i--)
    {
        if(stopthreads==0)
        {

            if(a3[i]==ele3)
            {
                pos3=i;
                stopthreads=1;
                break;
            }
        }
        else
        {
            Sleep(100);
            break;
        }
    }
    m.lock();
    cout<<"thread3 pos3 "<<pos3<<endl;
    m.unlock();
}

void linsearch4(int* a4,int ele4)
{
    int pos4=-1,i,mid;
    mutex m;
    mid = int(((n+1)/2)-1);
    for(i=n-1;i>=mid;i--)
    {
        if(stopthreads==0)
        {

            if(a4[i]==ele4)
            {
                pos4=i;
                stopthreads=1;
                break;
            }
        }
        else
        {   Sleep(100);
            break;
        }
    }
    cout<<"thread4 pos4 "<<pos4<<endl;
}